From 0df7010ce6ff72f80cdf8f8a3ce9bc58fb0225c4 Mon Sep 17 00:00:00 2001
From: Joe MacDonald <joe_macdonald@mentor.com>
Date: Thu, 24 Sep 2020 22:04:25 -0400
Subject: [PATCH 1/2] get_device: call get_partnum with dev, not devname

By the time get_partnum() is called here, devname may have been modified
to it no longer ends with the partition number but instead is a
null-terminated device name.  That is, dev may still be '/dev/sda3' but
devname is now '/dev/sda'.

Additionally, the '-n' parameter hasn't been respected, with get_partnum()
consistently returning '1' regardless of the actual partition specified on
the command line.

Only attempt to deduce the partition number if it wasn't already
explicitly stated on the command line.  This would also be essential if
mmc devices should be supported in the future as they're frequently named
mmcblk0p1 where a partition name isn't an int but a string.

Signed-off-by: Joe MacDonald <joe_macdonald@mentor.com>
---
 fatresize.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fatresize.c b/fatresize.c
index fe72c17..d85d03a 100644
--- a/fatresize.c
+++ b/fatresize.c
@@ -207,8 +207,8 @@ static int get_device(char *dev) {
       free(devname);
       return 0;
     }
-  } else {
-    opts.pnum = get_partnum(devname);
+  } else if (opts.pnum < 0) {
+    opts.pnum = get_partnum(dev);
   }
   ped_device_destroy(peddev);
   opts.device = devname;

From fee184f6c8a94628939fdca9b172cb711dbc2b41 Mon Sep 17 00:00:00 2001
From: Joe MacDonald <joe_macdonald@mentor.com>
Date: Thu, 24 Sep 2020 22:11:47 -0400
Subject: [PATCH 2/2] size: make a best effort to fulfill size request

Under certain circumstances we cannot accommodate the size request:  the
specified size is either greater than or exactly equal to the available
length.  This is very likely to happen if the user is specifying less
precise size values such as MiB, where the number argument is multiplied
by a factor or three.  Limit this to the actual partition length.

It also happens that if you specify the precise number of bytes free, you
also, you'll hit an assertion in ped_geometry_test_sector_inside() so
we'll bump the request down by one sector and run with that.

Signed-off-by: Joe MacDonald <joe_macdonald@mentor.com>
---
 fatresize.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/fatresize.c b/fatresize.c
index d85d03a..af66fd6 100644
--- a/fatresize.c
+++ b/fatresize.c
@@ -621,6 +621,11 @@ int main(int argc, char **argv) {
     opts.size = constraint->max_size * dev->sector_size;
     ped_constraint_destroy(constraint);
   }
+  if (opts.size >= dev->sector_size * part_geom.length) {
+     printf("Specified size (%llu) equals or exceeds partition size (%llu)\n",
+            opts.size, dev->sector_size * part_geom.length);
+     opts.size = (dev->sector_size * (part_geom.length - 1));
+  }
 
   start = part_geom.start;
   printd(3, "ped_geometry_new(%llu)\n", start);
